### DISCUSSION TITLE
[Proposal] Per-request state in SSR

### DISCUSSION THREAD

**Discussion by isaac-mcfadyen** [withastro/roadmap#221] [Proposal]
**The current problem:**
- currently, with Astro SSR you get the request object and can do stuff on it, including authentication
- but there's not really a way to store per-request authentication
- for example, platforms like Cloudflare Workers/Pages process other requests during an async task (such as a fetch to an external API provider)
- as you can imagine, this is disasterous because if you call an external API to authenticate a user and then another request comes in from a different user, things get mixed up

**(One) proposed solution:**
- "request state" (or whatever you want to call it).
- this would be state that is tied to a request. Platforms like SvelteKit already have this in the form of session and getSession
- if another request comes in, it would get a fresh "request state" (so the sessions would be tied to the request)
- this would avoid mess ups with concurrent requests by letting processing from one request only access it's own state

**Alternatives considered:**
- can't really think of many, on any platform like Cloudflare Workers/Pages that processes requests concurrently during async tasks you can easily end up with mixups between state

**Comment by isaac-mcfadyen** [comment]
Can't seem to set the category to RFC Discussions 🤔

**Comment by laptou** [comment]
This seems to be similar to what I proposed in #230. However, my design is not meant to allow you to set the per-request state from within Astro components, and is instead intended to allow you to create this state before Astro rendering begins and hand it off to the renderer. What do you think of this?

**Reply by isaac-mcfadyen** [reply]
The main issue with that is that with SSR platforms like Cloudflare Pages you don't have (much) control over the rendering itself (it's part of the plugin), you only have control over the components and framework.

The way SvelteKit does it is have a file called `hooks.ts` that they export an async hook that's called on render, where you can provide any special context (authentication, or you can actually run the authentication yourself in the hook). Maybe that would work?
